p2p/peersHolder/peersHolder.go

```diff
@@ -0,0 +1,199 @@
+package peersHolder
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go-core/core"
+)
+
+type peerInfo struct {
+	pid     core.PeerID
+	shardID uint32
+}
+
+type peerIDData struct {
+	connectionAddressSlice string
+	shardID                uint32
+	index                  int
+}
+
+type peersHolder struct {
+	preferredConnAddrSlices    []string
+	connAddrSliceToPeerInfo    map[string]*peerInfo
+	tempPeerIDsWaitingForShard map[core.PeerID]string
+	peerIDsPerShard            map[uint32][]core.PeerID
+	peerIDs                    map[core.PeerID]*peerIDData
+	sync.RWMutex
+}
+
+// NewPeersHolder returns a new instance of peersHolder
+func NewPeersHolder(preferredConnectionAddressSlices [][]byte) *peersHolder {
+	preferredConnections := make([]string, 0)
+	connAddrSliceToPeerIDs := make(map[string]*peerInfo)
+
+	for _, connAddrSlice := range preferredConnectionAddressSlices {
+		preferredConnections = append(preferredConnections, string(connAddrSlice))
```

Comment:
As discussed with @bogdan-rosianu we might have here a string validator for the input strings: let's permit only peer ids or valid ips. Nothing else. This string validator should be called only here, in the constructor as to be as light as possible in processing terms.